tests/msc2403_test.go

```diff
@@ -0,0 +1,281 @@
+				return true
+			})
+		})
+
+		t.Run("Users see state events from the room that they knocked on in /sync", func(t *testing.T) {
+			t.Parallel()
+			knockingUser.SyncUntil(
+				t,
+				"",
+				"rooms."+client.GjsonEscape("xyz.amorgan.knock")+"."+client.GjsonEscape(roomID)+".knock_state.events",
+				func(ev gjson.Result) bool {
+					// We don't current define any required state event types to be sent
+					// If we've reached this point, then an entry for this room was found
+					return true
+				},
+			)
+		})
+	})
+
+	if !federation {
+		// Rescinding a knock over federation is currently not supported in Synapse
+		// TODO: Preventing another homeserver from running this test just because Synapse cannot is unfortunate
+		t.Run("A user that has knocked on a local room can rescind their knock and then knock again", func(t *testing.T) {
+			// Rescind knock
+			knockingUser.MustDo(
+				t,
+				"POST",
+				[]string{"_matrix", "client", "r0", "rooms", roomID, "leave"},
+				struct {
+					Reason string `json:"reason"`
+				}{
+					"Just kidding!",
+				},
+			)
+
+			// Knock again
+			knockOnRoom(t, knockingUser, roomID, "Let me in... again?", []string{"hs1"})
+		})
+	}
+
+	t.Run("A user in the room can reject a knock", func(t *testing.T) {
+		// Reject the knock
+		inRoomUser.MustDo(
+			t,
+			"POST",
+			[]string{"_matrix", "client", "r0", "rooms", roomID, "kick"},
+			struct {
+				UserID string `json:"user_id"`
+				Reason string `json:"reason"`
+			}{
+				knockingUserID,
+				"I don't think so",
+			},
+		)
+
+		// Knock again
+		knockOnRoom(t, knockingUser, roomID, "Pleeease let me in?", []string{"hs1"})
+	})
+
+	t.Run("A user can knock on a room without a reason", func(t *testing.T) {
+		// Reject the knock
+		inRoomUser.MustDo(
+			t,
+			"POST",
+			[]string{"_matrix", "client", "r0", "rooms", roomID, "kick"},
+			struct {
+				UserID string `json:"user_id"`
+				Reason string `json:"reason"`
+			}{
+				knockingUserID,
+				"Please try again",
+			},
+		)
+
+		// Knock again, this time without a reason
+		knockingUser.MustDoRaw(
+			t,
+			"POST",
+			[]string{"_matrix", "client", "unstable", "xyz.amorgan.knock", roomID},
+			[]byte("{}"),
+			"application/json",
+			url.Values{"server_name": []string{"hs1"}},
+		)
+	})
+
+	t.Run("A user in the room can accept a knock", func(t *testing.T) {
+		inRoomUser.MustDo(
+			t,
+			"POST",
+			[]string{"_matrix", "client", "r0", "rooms", roomID, "invite"},
+			struct {
+				UserID string `json:"user_id"`
+				Reason string `json:"reason"`
+			}{
+				knockingUserID,
+				"Seems like a trustworthy fellow",
+			},
+		)
+	})
+
+	t.Run("A user cannot knock on a room they are already in", func(t *testing.T) {
+		reason := "I'm sticking my hand out the window and knocking again!"
+		knockOnRoomWithStatus(t, knockingUser, roomID, reason, []string{"hs1"}, 403)
+	})
+
+	t.Run("A user that is banned from a room cannot knock on it", func(t *testing.T) {
```

Comment:
Is this really testing what we think it is or is it testing that the user is already in the room? (Or does banning also kick the user?)

Reply:
Banning will kick a user, yes. It updates their membership to `ban`, removing them from the room.